src/Compilers/CSharp/Portable/Symbols/OverriddenOrHiddenMembersHelpers.cs

```diff
@@ -920,43 +905,89 @@ private static int CustomModifierCount(Symbol member)
         /// <summary>
         /// Determine if this method requires a methodimpl table entry to inform the runtime of the override relationship.
         /// </summary>
-        internal static bool RequiresExplicitOverride(this MethodSymbol method)
+        /// <param name="warnAmbiguous">True if we should produce an ambiguity warning per https://github.com/dotnet/roslyn/issues/45453 .</param>
+        internal static bool RequiresExplicitOverride(this MethodSymbol method, out bool warnAmbiguous)
+        {
+            warnAmbiguous = false;
+            if (!method.IsOverride)
+                return false;
+
+            MethodSymbol csharpOverriddenMethod = method.OverriddenMethod;
+            if (csharpOverriddenMethod is null)
+                return false;
+
+            MethodSymbol runtimeOverriddenMethod = method.GetFirstRuntimeOverriddenMethodIgnoringNewSlot(out bool wasAmbiguous);
+            if (csharpOverriddenMethod == runtimeOverriddenMethod && !wasAmbiguous)
+                return false;
+
+            // See https://github.com/dotnet/roslyn/issues/45453. No need to warn when the runtime
+            // supports covariant returns because the methodimpl is unambiguously understood by the runtime.
+            if (method.ContainingAssembly.RuntimeSupportsCovariantReturnsOfClasses)
+                return true;
+
+            // If the method was declared as a covariant return, there will be a compile-time error since the runtime
+            // does not support covariant returns. In this case we do not warn about runtime ambiguity and pretend that
+            // we can use a methodimpl (even though it is of a form not supported by the runtime and would result in a
+            // loader error) so that the symbol APIs produce the most useful result.
+            if (!method.ReturnType.Equals(csharpOverriddenMethod.ReturnType, TypeCompareKind.AllIgnoreOptions))
+                return true;
+
+            // Due to https://github.com/dotnet/runtime/issues/38119 the methodimpl would
+            // appear to the runtime to be ambiguous in some cases.
+            bool methodimplWouldBeAmbiguous = csharpOverriddenMethod.MethodHasRuntimeCollisionInSubstututedForm();
+            if (!methodimplWouldBeAmbiguous)
+                return true;
+
+            // We produce the warning when a methodimpl would be required but would be ambiguous to the runtime.
+            // This ambiguity only arises due to collision after generic substitution.
+            warnAmbiguous = csharpOverriddenMethod.ContainingType.IsGenericType;
```

Comment:
>csharpOverriddenMethod.ContainingType.IsGenericType [](start = 28, length = 51)

I am not sure if this check is accurate enough. If my understand is correct, this check is added to suppress confusing cascading warning for unit-test ```RecordTests.Overrides_01```. The type there is indeed not generic, but I think if we make it generic, we also don't want to report that warning. Also, I believe, ```IsGenericType``` is true when the type itself is not generic, but is contained in a generic type, I think we wouldn't want to report the warning for a non-generic type like that. I think several targeted tests are warranted around this issue. It looks like the usage of `records` feature isn't really required, it should be enough to declare the same virtual method twice in the same class and then attempt to override it. Also, consider if we could simply check whether both definitions of conflicting methods are completely equivalent and they are declared in source (i.e. in some compilation), which would constitute the error situation that compiler complains about elsewhere, and drop that check about generic types altogether. Please include a test for an attempt to override against retargeting methods, as apposed to an attempt to override against source methods directly.

Reply:
This is intended to be a simple check for one form of invalid program to suppress the cascaded warning. It isn't intended to suppress the warning in all possible scenarios where it could be considered cascaded. Given how rare this situation is likely to be in the first place, improvements to reduce the cascaded warnings would not be worth the effort.

---
In reply to: [449128610](https://github.com/dotnet/roslyn/pull/44025#discussion_r449128610) [](ancestors = 449128610)

Reply:
> This is intended to be a simple check for one form of invalid program to suppress the cascaded warning. It isn't intended to suppress the warning in all possible scenarios where it could be considered cascaded. Given how rare this situation is likely to be in the first place, improvements to reduce the cascaded warnings would not be worth the effort.

Could you please clarify what scenario you are trying to address? Since the commit adding this check didn't include any new tests, there is no certainty that my guess is accurate.

> The use of IsGenericType was intentional, as the return type may be defined using type parameters from an enclosing type.

Could you please elaborate. I am not sure why would this make any difference, Again, it is quite possible I misunderstand your intent here. 

> As you know, there is a bug in retargeted methods that would likely prevent a test of the form you suggest adding.

I didn't think the scenario is specific to Covariant Return scenarios. In that case, I am not aware of any bug that would prevent suggested testing. 

---
In reply to: [449160771](https://github.com/dotnet/roslyn/pull/44025#discussion_r449160771) [](ancestors = 449160771,449128610)

Reply:
Note, that my initial suggestion was about adjusting behavior of ```GetFirstRuntimeOverriddenMethodIgnoringNewSlot``` to not report an ambiguity in that particular error scenario and to undo the adjustment here. Only that method has access to both candidates. Sorry if that was not clearly expressed in the initial comment. 

---
In reply to: [449175779](https://github.com/dotnet/roslyn/pull/44025#discussion_r449175779) [](ancestors = 449175779,449160771,449128610)

Reply:
> Could you please clarify what scenario you are trying to address? Since the commit adding this check didn't include any new tests, there is no certainty that my guess is accurate.

I am trying to address the scenario that conflicting methods were declared in a non-generic type. There is an existing test that hits that scenario (inherited from master), which is why no additional test had to be added.

> Could you please elaborate? I am not sure why would this make any difference, Again, it is quite possible I misunderstand your intent here. 

A type that has no type parameters of its own but is nested inside a generic type can have a conflict in substituted form that would not have been a conflict where declared, due to the use of type parameters.

I don't consider it important which scenarios the warning is suppressed in, since in all applicable scenarios there is also an error reported. This change was mainly to simplify the merge with master.

---
In reply to: [449179890](https://github.com/dotnet/roslyn/pull/44025#discussion_r449179890) [](ancestors = 449179890,449175779,449160771,449128610)

Reply:
> Note, that my initial suggestion was about adjusting behavior of GetFirstRuntimeOverriddenMethodIgnoringNewSlot to not report an ambiguity in that particular error scenario and to undo the adjustment here. Only that method has access to both candidates. Sorry if that was not clearly expressed in the initial comment. 

The purpose of `GetFirstRuntimeOverriddenMethodIgnoringNewSlot` is to compute what the runtime overridden method would be, ignoring `newslot` and ignoring what the actual overridden method *should* be. As such, it has no business paying any attention to the proper (non-runtime) overridden member. It is used to compute whether or not we need to produce a `methodimpl`, which is what this method does, so it cannot use the `methodimpl` as part of that computation.

---
In reply to: [449185254](https://github.com/dotnet/roslyn/pull/44025#discussion_r449185254) [](ancestors = 449185254,449179890,449175779,449160771,449128610)

Reply:
> I am trying to address the scenario that conflicting methods were declared in a non-generic type. There is an existing test that hits that scenario (inherited from master), which is why no additional test had to be added.

Could you please point me to the exact unit-test?

> This change was mainly to simplify the merge with master.

I cannot say that I am comfortable with this approach. The merge from master indicated a possible flaw in the changes made in this PR. It is clear to me that more thorough testing is needed around this change, clearly demonstrating scenarios which you think is worth affecting and which you think is not worth affecting. Reviewers should have a clear understanding how this is going to affect different scenarios and should be able to form their own opinion about what is important and what is not important.  


---
In reply to: [449185254](https://github.com/dotnet/roslyn/pull/44025#discussion_r449185254) [](ancestors = 449185254,449179890,449175779,449160771,449128610)

Reply:
You can see the difference in the behavior of the test `RecordTests.Overrides_01`. We do not report the warning in that case.

If there is a specific flaw in the behavior of the compiler as implemented here, please identify it. All of the possible code paths are exercised by tests. There are certainly many other "scenarios" in which the warning may or may not be suppressed, but I don't believe it is important that any of them (or all of them) are suppressed.

---
In reply to: [449190145](https://github.com/dotnet/roslyn/pull/44025#discussion_r449190145) [](ancestors = 449190145,449185254,449179890,449175779,449160771,449128610)

Reply:
> An ambiguity there (in the runtime override) is irrelevant to whether or not the methodimpl would be ambiguous, as the methodimpl may well point to some other type altogether.

> The purpose of GetFirstRuntimeOverriddenMethodIgnoringNewSlot is to compute what the runtime overridden method would be, ignoring newslot and ignoring what the actual overridden method should be. As such, it has no business paying any attention to the proper (non-runtime) overridden member. It is used to compute whether or not we need to produce a methodimpl, which is what this method does, so it cannot use the methodimpl as part of that computation.

I am merely making a suggestion how to address an issue that in my opinion exists with the current code. If you think there is a better way to to improve things, that is great. There is really no point in arguing what should do what and what should pay attention to what. The resulting behavior is more important. We are in complete control and can redefine all these things. For example, you did that in this PR. It used to be that ```GetFirstRuntimeOverriddenMethodIgnoringNewSlot``` was not involved into deciding whether to report WRN_MultipleRuntimeOverrideMatches, but now it is involved and, I believe, that is one of the reasons why we started reporting WRN_MultipleRuntimeOverrideMatches in cases when we didn't before.    

---
In reply to: [449190145](https://github.com/dotnet/roslyn/pull/44025#discussion_r449190145) [](ancestors = 449190145,449185254,449179890,449175779,449160771,449128610)

Reply:
> You can see the difference in the behavior of the test RecordTests.Overrides_01. We do not report the warning in that case.

> If there is a specific flaw in the behavior of the compiler as implemented here, please identify it. All of the possible code paths are exercised by tests. There are certainly many other "scenarios" in which the warning may or may not be suppressed, but I don't believe it is important that any of them (or all of them) are suppressed.

Ok, let me do my own testing then.

---
In reply to: [449195217](https://github.com/dotnet/roslyn/pull/44025#discussion_r449195217) [](ancestors = 449195217,449190145,449185254,449179890,449175779,449160771,449128610)

Reply:
> Ok, let me do my own testing then.

With the latest state of the branch (commit 41) I simply tested scenarios that were outlined at the beginning of this thread. I.e. conflicting methods declared in a generic type and conflicted methods declared in a non-generic type, which is nested into a generic type. These were compared against the scenario when conflicted methods were declared in a non-generic top-level type. I also paid attention to behavior inconsistencies between targeting runtime that supports Covariant Returns feature and runtime that doesn't support it.

My expectations were:
- There should be no difference in behavior across different flavors of type genericity, because it really makes no difference since methods conflict at their original definition point
- There should be no difference in behavior across different flavors of targeted runtimes, because the scenario doesn't involve the Covariant Returns feature, nor could it benefit from any fix that a runtime supporting the feature might have.
- There should be no WRN_MultipleRuntimeOverrideMatches warnings reported for scenarios for which this warning is not reported in master branch right now.

The full code of the unit-tests is below, they reflect behavior of the branch at commit 41.
Here is the summary:
1) Test03 - a non-generic top-level type
-  WRN_MultipleRuntimeOverrideMatches warning is not reported, it is not reported for this scenario in master.
-  Values of IsMetadataNewSlot and RequiresExplicitOverride on the override flip-flop depending on whether runtime supports Covariant Returns feature.
2) Test04 - a generic top-level type
-  Presence of the WRN_MultipleRuntimeOverrideMatches warning flip-flops depending on whether runtime supports Covariant Returns feature.
-  Values of IsMetadataNewSlot and RequiresExplicitOverride on the override flip-flop depending on whether runtime supports Covariant Returns feature.
3) Test05 - a non-generic type nested into generic type
-  Presence of the WRN_MultipleRuntimeOverrideMatches warning flip-flops depending on whether runtime supports Covariant Returns feature. It is not reported for this scenario in master.
-  Values of IsMetadataNewSlot and RequiresExplicitOverride on the override flip-flop depending on whether runtime supports Covariant Returns feature.

I don't believe there is a benefit in this "erratic" behavior across scenarios.

I also believe that the change
```
            // We produce the warning when a methodimpl would be required but would be ambiguous to the runtime.
            // This ambiguity only arises due to collision after generic substitution.
            warnAmbiguous = csharpOverriddenMethod.ContainingType.IsGenericType;
```
is confusing to the reader of the code in the sense that it doesn't go to the core of the issue we want to handle, which was indicated by the failing tests that came from master. Yes, the change in commit 41 suppresses the warning for some specific scenario, but it is not general enough to cover all family of them. Besides that, we still have the issue with flip-flopping IsMetadataNewSlot and RequiresExplicitOverride API across all scenarios.

I would like to propose a simple fix to address all these problems together. Instead of making changes in ```RequiresExplicitOverride```, I suggest to make a small change in ```GetFirstRuntimeOverriddenMethodIgnoringNewSlot``` function. Right before we record the fact of ambiguity, we could add a check that explicitly targets duplicate method declarations:
```
                                if (MemberSignatureComparer.RuntimeSignatureComparer.Equals(candidate.OriginalDefinition, overridden.OriginalDefinition))
                                {
                                    // Identical definitions in the original type definitions. This is not supported by any runtime and it is probably 
                                    // better to not treat this as an ambiguity, since an error is likely has been reported elsewhere.
                                    continue;
                                }

                                // found more than one possible override in this type
                                wasAmbiguous = true;
                                return candidate;
```

After this change all three tests start behaving as I would expect them to behave. I even ran all Roslyn tests, all of them passed.
Given the simplicity of the proposed change, focus on specific condition that causes the trouble, and the consistency benefits that it gives, I believe it is a better option by comparison to the current approach.  

The tests:
```
        [Theory]
        [CombinatorialData]
        public void Test_03(bool withCovariantCapableRuntime)
        {
            var src = @"
class Base
{
    public virtual void M1() => throw null; // 1

    public virtual void M1() => throw null; // 2
}

class Derived : Base
{
    public override void M1() => throw null; // 3
}
";

            var comp = CreateCompilationWithCovariantReturns(src, pretendRuntimeSupportsCovariantReturnsFeature: withCovariantCapableRuntime);
            var symbol = comp.GetMember<MethodSymbol>("Derived.M1");
            Assert.Equal("void Derived.M1()", symbol.ToTestDisplayString());

            comp.VerifyDiagnostics(
                // (6,25): error CS0111: Type 'Base' already defines a member called 'M1' with the same parameter types
                //     public virtual void M1() => throw null; // 2
                Diagnostic(ErrorCode.ERR_MemberAlreadyExists, "M1").WithArguments("M1", "Base").WithLocation(6, 25)
                );

            Assert.Equal(withCovariantCapableRuntime, symbol.IsMetadataNewSlot()); // Expected False regardless of runtime
            Assert.Equal(withCovariantCapableRuntime, symbol.RequiresExplicitOverride(out _)); // Expected False regardless of runtime
        }

        [Theory]
        [CombinatorialData]
        public void Test_04(bool withCovariantCapableRuntime)
        {
            var src = @"
class Base<T>
{
    public virtual void M1() => throw null; // 1

    public virtual void M1() => throw null; // 2
}

class Derived : Base<int>
{
    public override void M1() => throw null; // 3
}
";

            var comp = CreateCompilationWithCovariantReturns(src, pretendRuntimeSupportsCovariantReturnsFeature: withCovariantCapableRuntime);

            var symbol = comp.GetMember<MethodSymbol>("Derived.M1");
            Assert.Equal("void Derived.M1()", symbol.ToTestDisplayString());

            // (6,25): error CS0111: Type 'Base<T>' already defines a member called 'M1' with the same parameter types
            //     public virtual void M1() => throw null; // 2
            var common1 = Diagnostic(ErrorCode.ERR_MemberAlreadyExists, "M1").WithArguments("M1", "Base<T>").WithLocation(6, 25);
            // (11,26): error CS0462: The inherited members 'Base<T>.M1()' and 'Base<T>.M1()' have the same signature in type 'Derived', so they cannot be overridden
            //     public override void M1() => throw null; // 3
            var common2 = Diagnostic(ErrorCode.ERR_AmbigOverride, "M1").WithArguments("Base<T>.M1()", "Base<T>.M1()", "Derived").WithLocation(11, 26);

            if (withCovariantCapableRuntime)
            {
                comp.VerifyDiagnostics(common1, common2);
            }
            else
            {
                comp.VerifyDiagnostics(
                    // (4,25): warning CS1957: Member 'Derived.M1()' overrides 'Base<int>.M1()'. There are multiple override candidates at run-time. It is implementation dependent which method will be called.
                    //     public virtual void M1() => throw null; // 1
                    Diagnostic(ErrorCode.WRN_MultipleRuntimeOverrideMatches, "M1").WithArguments("Base<int>.M1()", "Derived.M1()").WithLocation(4, 25),
                    common1, common2
                    );
            }

            Assert.Equal(withCovariantCapableRuntime, symbol.IsMetadataNewSlot()); // Expected False regardless of runtime
            Assert.Equal(withCovariantCapableRuntime, symbol.RequiresExplicitOverride(out _)); // Expected False regardless of runtime
        }

        [Theory]
        [CombinatorialData]
        public void Test_05(bool withCovariantCapableRuntime)
        {
            var src = @"
class Container<T>
{
    class Base
    {
        public virtual void M1() => throw null; // 1

        public virtual void M1() => throw null; // 2
    }

    class Derived : Base
    {
        public override void M1() => throw null; // 3
    }
}
";

            var comp = CreateCompilationWithCovariantReturns(src, pretendRuntimeSupportsCovariantReturnsFeature: withCovariantCapableRuntime);

            var symbol = comp.GetMember<MethodSymbol>("Container.Derived.M1");
            Assert.Equal("void Container<T>.Derived.M1()", symbol.ToTestDisplayString());

            // (8,29): error CS0111: Type 'Container<T>.Base' already defines a member called 'M1' with the same parameter types
            //         public virtual void M1() => throw null; // 2
            var common = Diagnostic(ErrorCode.ERR_MemberAlreadyExists, "M1").WithArguments("M1", "Container<T>.Base").WithLocation(8, 29);

            if (withCovariantCapableRuntime)
            {
                comp.VerifyDiagnostics(common);
            }
            else
            {
                comp.VerifyDiagnostics(
                    // (6,29): warning CS1957: Member 'Container<T>.Derived.M1()' overrides 'Container<T>.Base.M1()'. There are multiple override candidates at run-time. It is implementation dependent which method will be called. Please use a newer runtime.
                    //         public virtual void M1() => throw null; // 1
                    Diagnostic(ErrorCode.WRN_MultipleRuntimeOverrideMatches, "M1").WithArguments("Container<T>.Base.M1()", "Container<T>.Derived.M1()").WithLocation(6, 29),
                    common
                    );
            }

            Assert.Equal(withCovariantCapableRuntime, symbol.IsMetadataNewSlot()); // Expected False regardless of runtime
            Assert.Equal(withCovariantCapableRuntime, symbol.RequiresExplicitOverride(out _)); // Expected False regardless of runtime
        }
```


   

---
In reply to: [449196034](https://github.com/dotnet/roslyn/pull/44025#discussion_r449196034) [](ancestors = 449196034,449195217,449190145,449185254,449179890,449175779,449160771,449128610)

Reply:
I have recorded [https://github.com/dotnet/roslyn/issues/45678](https://github.com/dotnet/roslyn/issues/45678) for the suggestion to reduce cascaded diagnostics.

The behavior of the compiler internal methods `IsMetadataNewSlot()` and `RequiresExplicitOverride()` is intended. The values of these internal methods are expected to "flip-flop" depending on whether or not the target runtime contains a fix for [https://github.com/dotnet/runtime/issues/38119](https://github.com/dotnet/runtime/issues/38119), as these methods are used to implement the compat council's decision documented in [https://github.com/dotnet/roslyn/issues/45453](https://github.com/dotnet/roslyn/issues/45453). There are many tests in `CovariantReturnTests` that illustrate this behavior. Being internal methods, this behavior is not directly observable by consumers of the compiler.

---
In reply to: [449346129](https://github.com/dotnet/roslyn/pull/44025#discussion_r449346129) [](ancestors = 449346129,449196034,449195217,449190145,449185254,449179890,449175779,449160771,449128610)

Reply:
> The behavior of the compiler internal methods IsMetadataNewSlot() and RequiresExplicitOverride() is intended. The values of these internal methods are expected to "flip-flop" depending on whether or not the target runtime contains a fix for [https://github.com/dotnet/runtime/issues/38119](https://github.com/dotnet/runtime/issues/38119), as these methods are used to implement the compat council's decision documented in [https://github.com/dotnet/roslyn/issues/45453](https://github.com/dotnet/roslyn/issues/45453).

I do not agree with your interpretation of compat council decision. I believe it only applies to scenarios that are really affected by the runtime fix, and is focused only on preserving backward compatibility when compiler emits code against runtimes that do not have the fix. None of the scenarios discussed on this thread are applicable on either point: they are not affected by the runtime fix, there is no codegen to preserve  because no code is generated. Finally, I do not believe there is any benefit from the warning or APIs flip flopping for these scenarios. In my opinion, the logic for detecting ambiguity was moved to ```GetFirstRuntimeOverriddenMethodIgnoringNewSlot ``` with the intent to implement compat council's decision, but that was done in a way that doesn't accurately match the original logic, thus causing the warning to be reported for scenarios for which it wasn't reported before. I think, compat council never expressed a desire to change compiler's behavior in this way. 
Nevertheless, I am fine with following up on this separately, under umbrella of the opened issue.  

---
In reply to: [450310947](https://github.com/dotnet/roslyn/pull/44025#discussion_r450310947) [](ancestors = 450310947,449346129,449196034,449195217,449190145,449185254,449179890,449175779,449160771,449128610)